packages/flutter/lib/src/widgets/magnifier.dart

```diff
@@ -0,0 +1,175 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+
+/// A widget that magnifies a screen region relative to itself.
+///
+/// [Magnifier] may have a [child], which will be drawn over the lens. This is useful
+/// for overlays, like tinting the lens.
+///
+/// Some caveats for using the magnifier:
+/// * [Magnifier] may only display widgets that come before it in the paint order; for example,
+/// if magnifier comes before `widget A` in a column, then you will not be able to see `widget A`
+/// in the magnifier.
+/// *  If the magnifier points out of the bounds of the app, will have undefined behavior. This 
+/// generally results in the magnifier having undesired transparency, i.e. showing the layers
+/// underneath it. 
+///
+///
+/// This widget's magnification does not lower resolution of the subject
+/// in the [Magnifier].
+///
+/// See also:
+/// * [BackdropFilter], which [Magnifier] uses along with [ImageFilter.matrix] to
+/// Magnify a screen region.
+/// * [Loupe], which uses [Magnifier] to magnify text.
+class Magnifier extends SingleChildRenderObjectWidget {
+  /// Construct a [Magnifier],
+  Magnifier(
+      {super.key,
+      super.child,
+      ShapeBorder? shape,
+      this.magnificationScale = 1,
+      this.focalPoint = Offset.zero})
+      : clip = shape != null
+            ? ShapeBorderClipper(
+                shape: shape,
+              )
+            : null;
+
+  ///  [focalPoint] of the magnifier is the area the center of the
+  /// [Magnifier] points to, relative to the center of the magnifier.
+  /// If left as [Offset.zero], the magnifier will magnify whatever is directly
+  /// below it.
+  final Offset focalPoint;
+
+  /// The scale of the magnification.
+  ///
+  /// A [magnificationScale] of 1 means that the content magi
+  final double magnificationScale;
+
+  /// The shape of the magnifier is dictated by [clip], which clips
+  /// the magnifier to the shape. If null, the shape will be rectangular.
+  final ShapeBorderClipper? clip;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return _RenderMagnification(focalPoint, magnificationScale, clip);
+  }
+
+  @override
+  void updateRenderObject(
+      BuildContext context, covariant RenderProxyBox renderObject) {
+    (renderObject as _RenderMagnification)
```

Comment:
Since `Magnifier` should be public, but `_RenderMagnification` should not be, this will trigger the "no private members in public API lint`. To workaround, cast the renderObject to it's private subtype.  

[Prior art](https://github.com/flutter/flutter/blob/cb9a1d623ccd460c3744ed019af76496c47a353e/packages/flutter/lib/src/material/mergeable_material.dart#L682-L688)